src/main/java/com/google/sps/model/GmailResponse.java

```diff
@@ -0,0 +1,288 @@
+        if (timestamp > mostRecentEmailTimestamp) {
+          this.mostRecentEmailTimestamp = timestamp;
+        }
+      }
+
+      public void incrementTimesSent() {
+        timesSent++;
+      }
+    }
+
+    // If there are no emails, return "" (the default value for sender)
+    if (unreadEmails.isEmpty()) {
+      return "";
+    }
+
+    // Find out which sender sent the most unread emails.
+    // This is extracted from the "From" headers within messages (which are always present in a
+    // valid email)
+    // Store the senders of emails in a hashmap paired with 1) how often they sent emails in the
+    // last nDays and 2) what the timestamp was of their most recent email.
+    HashMap<String, EmailFrequencyWithDate> senders = new HashMap<>();
+    unreadEmails.forEach(
+        (Message m) -> {
+          long timestamp = m.getInternalDate();
+          MessagePart messagePayload = m.getPayload();
+
+          if (messagePayload == null || timestamp == 0) {
+            throw new GmailMessageFormatException("Messages must be of format METADATA or FULL");
+          }
+
+          String sender =
+              messagePayload.getHeaders().stream()
+                  .filter((MessagePartHeader header) -> header.getName().equals("From"))
+                  .findFirst()
+                  .get()
+                  .getValue();
+
+          if (senders.get(sender) == null) {
+            EmailFrequencyWithDate newEmailFrequencyWithDate =
+                new EmailFrequencyWithDate(timestamp);
+            senders.put(sender, newEmailFrequencyWithDate);
+          } else {
+            EmailFrequencyWithDate updatedEmailFrequencyWithDate = senders.get(sender);
+            updatedEmailFrequencyWithDate.incrementTimesSent();
+            updatedEmailFrequencyWithDate.updateMostRecentEmailTimestamp(timestamp);
+            senders.replace(sender, updatedEmailFrequencyWithDate);
+          }
+        });
+
+    // Senders cannot be empty, as all emails have a "From" header and
+    // there is at least one email in unreadEmails.
+    // Identify the sender who sent the most emails. In the case of a tie,
+    // choose the sender who sent an email most recently. In case of another tie
+    // (i.e. two senders with same send frequency and same timestamp),
+    // return either sender (this case doesn't need to be handled)
+    String headerValue =
+        senders.entrySet().stream()
+            .reduce(
+                (a, b) -> {
+                  if (a.getValue().getTimesSent() > b.getValue().getTimesSent()) {
+                    return a;
+                  } else if (b.getValue().getTimesSent() > a.getValue().getTimesSent()) {
+                    return b;
+                  } else {
+                    if (a.getValue().getMostRecentEmailTimestamp()
+                        > b.getValue().getMostRecentEmailTimestamp()) {
+                      return a;
+                    }
+                    return b;
+                  }
```

Comment:
Try to reduce branching. Maybe something like the following could be a bit cleaner:
```
if (a.getValue().getTimesSent() != b.getValue().getTimesSent()) {
  return a.getValue().getTimesSent() > b.getValue().getTimesSent() ? a : b;
}
return a.getValue().getMostRecentEmailTimestamp() > b.getValue().getMostRecentEmailTimestamp() ? a : b;
```

Reply:
Thanks for this! I ended up refactoring this method, but I'll keep this in mind for the future